lib/internal/modules/esm/get_format.js

```diff
@@ -91,7 +91,7 @@ function underNodeModules(url) {
 let typelessPackageJsonFilesWarnedAbout;
 function warnTypelessPackageJsonFile(pjsonPath, url) {
   typelessPackageJsonFilesWarnedAbout ??= new SafeSet();
-  if (!typelessPackageJsonFilesWarnedAbout.has(pjsonPath)) {
+  if (!underNodeModules(url) && !typelessPackageJsonFilesWarnedAbout.has(pjsonPath)) {
```

Comment:
One thing I noticed is that packages like `xstate` might have module pointing to a `.js` file while their package.json does not contain the `"type"` field (instead, they are using the top-level `"module"` field recognized by the bundlers). That might be the reason why @targos thought there was an error (because of the warnings?). I think for better UX in loading these modules, we should just avoid emitting the warning for typeless package.json for files under node_modules, which is a good practice in general.

Reply:
What are the current preconditions for the warning and what the warning is? (posted from mobile so I can't easily navigate the code here)

Reply:
If the package.json doesn't contain "type", and the file is `.js`, and it's ESM, Node.js would try to parse that first as CommonJS - if it throws a SyntaxError that makes it look like ESM, then Node.js tries to reparse it as ESM. The warning is there to prevent the situation where you have thousands of your files parsed twice instead of just once, and remind you you can avoid this if you just add `type` to your package.json.

Reply:
Oh, so it has to be reparsed, hmm. Couldn't the `module` files always be treated as ESM right from the start? I don't know of any package that would ship CJS in there.

> and remind you you can avoid this if you just add type to your package.json.

In the case of XState, I'd have to declare `type: "commonjs"` and that wouldn't really help anything here, right? So the best solution, really, would be to point `module` condition to an `.mjs` file.

Reply:
> Couldn't the module files always be treated as ESM right from the start? I don't know of any package that would ship CJS in there.

This reparse only happens in the absence of a "type" field in package.json. If there is already `"type": "module"` then all .js files would be treated as ESM directly, and if there is `"type": "commonjs"` then .js files would be treated as CommonJS (and won't get a reparse)

> In the case of XState, I'd have to declare type: "commonjs" and that wouldn't really help anything here, right?

That would prevent a reparse, so it would load a bit faster :) But if it's governed by `"type": "commonjs"`, currently this would require the `.mjs` extension to indicate an exception in the format of the file. I wonder if we should allow the `"module": "/path/to/file.js"` to override `"type": "commonjs"` and indicate that this `.js` file is ESM, that seems like an intuitive thing to do in this case?

Reply:
We are not using an explicit `"type": "commonjs"` since it feels rather reasonable to just rely on it being the default.

> and if there is "type": "commonjs" then .js files would be treated as CommonJS (and won't get a reparse)

I'm not sure if I got your whole intention right but it feels to me extremely weird that reparsing of `"module": "some-file.js"` would be dependent on `{ "type": "commonjs" }` vs `{}`. I understand why you wouldn't reparse at all with `{ "type": "module" }`.

> I wonder if we should allow the "module": "/path/to/file.js" to override "type": "commonjs" and indicate that this .js file is ESM, that seems like an intuitive thing to do in this case?

This seems reasonable to me. The whole point of supporting the module condition is to allow ESM contained by it to be consumed by both loaders. I don't see why I'd like to ever have that file to reference CJS file and I think it would be better to treat those as ESM right from the start. Most people consider a footgun that an `"import"` can legally reference a CJS file and that the condition doesn't imply the module format of the referenced file. Maybe, from that PoV, it would be inconsistent to treat `"module"` as ESM unconditionally (even if it would match the people expectations). It feels more like a theoretical than a practical concern though.

> That would prevent a reparse, so it would load a bit faster :)

That being said... I'll adjust to whatever you decide here. If I'll have to update all of my `"module": "*.js"` to `"module": "*.mjs"`, I'm fine with it.

Reply:
> I'm not sure if I got your whole intention right but it feels to me extremely weird that reparsing of "module": "some-file.js" would be dependent on { "type": "commonjs" } vs {}

The reparsing logic goes like this:

1. If "type" is "commonjs", then all .js files are treated as CommonJS explicitly - if they contain ESM syntax, then it would be regarded as a mistake (this has been the behavior ever since `type` was introduced)
2. If "type" is "module", then all .js files are treated as ESM explicitly. If they rely on a global `require()` or other CommonJS-only globals, then it would be regarded as a mistake (this has been the behavior ever since `type` was introduced)
3. If "type" is not undefined in any upper-level package.json, then `.js` files are treated as being ambiguous about its module type. Up until 22.7.0, Node.js treated them as CommonJS for backwards compatibility and if they use ESM syntax, Node.js would throw. After https://github.com/nodejs/node/pull/53619 (which has been shipped since 22.7.0), Node.js would attempt to reparse them as ESM if the syntax error indicates that they look like ESM and would accept them as ESM if the second parse succeeds.

So for this particular PR here, the question is whether we should treat the additional clue of an exports condition `"module": "/path/to/file.js"` as a new hint of the module type. But 1-3 would remain unaffected by this PR.

Reply:
Thanks for the explanation. I wasn't aware of the changes made to 3.

> IMO this feels the most natural

Same 👍

Reply:
It's worth noting that while it seems like it might make things work, conditions are not supposed to alter interpretation. Using them as an interpretation signal risks creating some confusing bugs where eg refactoring a file to no longer be an entry point changes the way it's interpreted. I would therefore advise against this approach in this case, to rather stick to the well-defined interpretation rules.

Reply:
Couldn’t the same thing (refactoring hazard) be said about moving files around ur project when u have directories with `package.json` scopes? One directory might be configured to have ESM .js files and another might be configured to have CJS .js files.

Personally, I’m not convinced by the above argument. Those rules might be well-defined today but they are certainly not exactly well-understood by the community at large.

Reply:
I wouldn't recommend directories with lots of package.json scopes either tbh.

Reply:
> Using them as an interpretation signal risks creating some confusing bugs where eg refactoring a file to no longer be an entry point changes the way it's interpreted

I think we can always go back on this when that happens, it's going to be experimental anyway, and the initial behavior should be as intuitive as possible even with a minor risk. If we are talking about asking 10% of the people doing `"module" : "/path/to/file.js"` conditions with an undefined `package.json#type`  v.s. risking someone seeing an error when they move the `"/path/to/file.js"` to be no longer an entrypoint with an undefined `package.json#type` , then the first one is worth the trade-off to me. The worst that could happen is that users would see an error telling them to update `package.json#type`, which they should've done already, we were only making a best-effort to save them the trouble, and give up when it's too hard. This goes along with our general trend of trying to make module authoring as intuitive as possible e.g. adding module format detection.

Reply:
Yes, but we shouldn't compromise on the resolver to support legacy code as well either. Many tools need to implement the Node.js resolver and this new rule will add many new lines to the specification for the resolver massively increasing complexity for an already complex resolver. That complexity must be justified - we cannot make arguments the reverse, to justify removing things that aren't needed.

If this work goes ahead then:

1. It must be a well-defined algorithm performing a backwards lookup on the conditions. If something appears as a single wildcard `#import` condition once, does that act as act as a signal? I'd think so otherwise there'd be ambiguity... so it needs to deconstruct over the entire resolultion space
2. It must be specified in the Node.js resolver as part of the DETECT_ESM_FORMAT_RULES.

Again, I will advise that the above is complex and not complexity Node.js needs, but (1) and (2) would be bare minimum requirements for this to go forward. Happy to discuss further.

Reply:
> Many tools need to implement the Node.js resolver and this new rule will add many new lines to the specification for the resolver massively increasing complexity for an already complex resolver.

They already need to do this after we have shipped module detection. This new condition merely fast-fowards the detection for ESM (or I guess for them, it can be used to cut corners to not do detection at all, it's up to them).

Reply:
Consider the following package.json:

```js
{
  "name": "pkg",
  "exports": {
    "main": "./index.js"
    "./submodules/x/*": {
      "module": "./submodules/x/*.js"
    }
  }
}
```

In the above all of `submodules/x/*.js` are set to `module` conditions so should be interepreted as ESM.

So if I do `import 'pkg'` this should also trigger the "module as ESM" path rule in the resolver. Or if not, we'd need to strictly define why wildcards do not get this support. Also there is imports and nested conditions.

This "reverse lookup" for exports and imports is what I am referring to as new resolver complexity that should be justified based on benefit for increasing resolver complexity. Regardless, it must be specified in the ESM resolver to land so other tools can follow hence requirement (1) above.

Reply:
> 1. It must be a well-defined algorithm performing a backwards lookup on the conditions.
> 2. It must be specified in the Node.js resolver as part of the DETECT_ESM_FORMAT_RULES.

I think those are looking at the wrong places - it does not need a backward lookup, but instead keeping track of the conditions already applied, and in `PACKAGE_EXPORTS_RESOLVE`, if the conditions applied includes `"module"`, override the format as `module`.

Reply:
@joyeecheung the issue with that approach is having a different interpretation based on whether you import `pkg/submodules/x/dep.js` directly or whether you import `pkg` and get it indirectly through the direct dependency import. So there is a race on the interpretation of the module - it has an ambiguous format depending on how it is imported. This is something we worked very hard to avoid in the ESM resolver.

Reply:
> but import 'pkg' won't lead to resolving import './submodules/x/dep.js through package.json#exports so it won't ever know that module condition was applied to it

Yes, but it will still be loaded as ESM later due to detection. It's just going to be parsed as CommonJS first, which is a bit silly, and is why we advise to use the type field to skip this futile attempt.

> having a different interpretation based on whether you import pkg/submodules/x/dep.js directly or whether you import pkg and get it indirectly through the direct dependency import

I think in reality, the interpretation would still be the same in most cases due to detection. If it's ESM and you don't have an explicit type field, then it will still be interpreted as ESM after the reparse. And as mentioned, the module condition only fast-forwards the detection.

Reply:
Anyway, I don't think the fast-forwarding must be implemented as part of this PR. It just feels silly to not use this to fast forward detection, but for most use cases, it's just a difference between seeing a warning + slowdown (current) v.s. not seeing a warning + not slowing down (what I am proposing to add), the interpretation of the module is likely to be still the same due to detection.

Reply:
Right, but we worked hard to ensure there is a unique module format for each module in the module system (`ESM_MODULE_FORMAT`). You're now saying that this function will no longer work correctly. That is not a decision to be taken lightly.

Reply:
IMO, theocratical purity should not be prioritized over user experience, the goal is just to not make shipping ESM harder than it already is. If a less pure spec makes the shipping experience nicer, then it's worth it.

By the way, the detection already throws in a `_source_` out of nowhere in `ESM_FILE_FORMAT`, with no explanation about how that format-deciding source is obtained, technically that can be a changing variable that varies depends on how the module is loaded (which it is, if you use the loader hooks), so the loophole has already been there.

Reply:
@joyeecheung invariants are important, especially when bundlers need to reimplement our algorithms. They should not be thrown away without careful discussion. Let's have that discussion certainly, and I'm happy to do that.

Reply:
It feels like a bummer that the user experience of running unbundled ESM in Node.js needs to give way to the developer experience of bundlers currently in use, when for most people, not needing to use any bundlers to run their code in Node.js should yield a better user experience . But that is a bit off topic already. I will leave the ESM loading optimization out of this PR.